scope::resolve throws more meaningful exceptions for statuses such as instructionLimitExceeded

michaelDCurran opened the pull request:
Fixes #87

The following new exceptions are thrown by UiaOperationScope::Resolve due to their respective operation result status values:
* ExecutionFailureException
* MalformedBytecodeException
* UnhandledRemoteException
* InstructionLimitExceededException

These new exceptions all inherit from winrt::hresult_error, so existing code that is catching winrt::result_error will still function exactly the same way.

Further to this, if InstructionLimitExceededException is thrown, any result variables already calculated before the limit was hit will be still resolved, making them available locally. 

These changes make it technically possible now to write remoted algorithms with the abstraction library, such that they could be continued on in a second remote call, if the first failed with an InstructionLimitExceeded exception, as result variables could be used to keep track of the state of the algorithm.
Also, all changes have been made to be fully compatible with previous code.

In order to make all this work, several other changes needed to be made.

The low-level winrt API's AutomationRemoteOperationResultSet::OperationStatus method was problematic as it essentially just returned the ExtendedError value, but tweaked for certain status values. E.g. E_FAIL was returned for InstructionLimitExceeded. This then meant that higher-level APIs could never access the actual OperationStatus value.
Therefore, the AutomationRemoteOperationResultSet class has the following changes:
* Added a Status method, which returns the raw AutomationRemoteOperationStatus value of the result. E.g. success, MalformedBytecodeInstruction, UnhandledException, ExecutionFailure. 
* Added an ExtendedError method, which returns the ExtendedError HRESULT value.
* Added a comment to the OperationStatus method noting it should be deprecated in favor of the separate Status and ExtendedError methods.
 
The high-level abstraction library's UiaOperationScope::ResultInternal now fetches Status and ExtendedError separately, rather than calling the old OperationStatus method. It then gives both of these values back by reference.
However, to ensure compatibility for higher-level APIs, extendedError is forced to E_FAIL if the status is InstructionLimitExceeded. Note that not doing this, ExtendedError would be S_OK. 

ResultInternal also now resolves all available result variables when the status is InstructionLimitExceeded, not just when the status is Success.

The high-level abstraction library's UiaOperationScope::ResolveHr method now returns the ExtendedError from ResolveInternal. But from the caller's point of view, there is no change at all.
 
The high-level abstraction library's UiaOperationScope::Resolve method now calls ResolveInternal directly, rather than ResolveHr. This is so it gets access to both the raw status and ExtendedError.
Based on the Status, Resolve throws the appropriate Exception, instantiated with the ExtendedError as its HRESULT error code.

Finally, tests have been added:
* executionFailureException: Testing that this exception is thrown by trying to execute a remote operation interacting with elements from more than one process. 
* UnhandledRemoteException: Testing that this exception is thrown by trying to fetch the first element from an empty UiaArray.
* InstructionLimitExceededException: Testing that this exception is thrown by constantly incrementing a UiaInt until the instruction limit is hit. 

I was not able to come up with any way of testing MalformedBytecodeException - the winrt and abstraction APIs should never allow a caller to produce this situation anyway I think.

michaelDCurran committed:
* AutomationRemoteOperationResultSet: add status and extendedError methods, and deprecate OperationStatus method.
* UIA operation abstraction library: scope::Resolve now throws subclasses of wil::ResultException based on operation status. I.e. ExecutionFailureException, InstructionLimitExceededException, MalformedBytecodeException, UnhandledRemoteException.

beweedon commented on src/UIAutomation/UiaOperationAbstraction/UiaOperationAbstraction.cpp:
```diff
@@ -2316,22 +2316,37 @@ namespace UiaOperationAbstraction
+        switch(status)
+        {
+            case winrt::Windows::UI::UIAutomation::Core::AutomationRemoteOperationStatus::MalformedBytecode:
+            throw MalformedBytecodeException(extendedError);
```
**nit:** Could we indent the `case` bodies one more?

beweedon commented on src/UIAutomation/UiaOperationAbstraction/UiaOperationAbstraction.cpp:
```diff
@@ -2316,22 +2316,37 @@ namespace UiaOperationAbstraction
     CATCH_RETURN();
   
-    void UiaOperationScope::ResolveHrInternal(HRESULT& operationResult)
+    void UiaOperationScope::ResolveInternal(winrt::Windows::UI::UIAutomation::Core::AutomationRemoteOperationStatus& status, HRESULT& extendedError)
```
While we're here could we change this to return an `std::pair<AutomationRemoteOperationState, HRESULT>` instead of using out params?

beweedon commented on src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp:
```diff
@@ -2715,5 +2715,113 @@ namespace UiaOperationAbstractionTests
         }
+
+        // Test that hitting the instruction limit in a remote operation causes an InstructionLimitExceededException in resolve.
+        TEST_METHOD(instructionLimitExceeded)
```
**nit:** Capitalize test names

beweedon commented on src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp:
```diff
@@ -2715,5 +2715,113 @@ namespace UiaOperationAbstractionTests
+                    throw;
+                }
+            });
+        }
```
We can produce malformed bytecode errors by passing over 1MB of bytecode to the remote process. Could we add a test for that? It should be possible to do that similarly to the instruction limit test, but with an actual C++ loop:

```cpp
for (int i = 0; i < 1048576; ++i)
{
    UiaBool b = true;
}
```

We definitely don't need to iterate this many times (the `NewBool` instruction plus its operands is a total of 9 bytes), but this is the basic idea.

beweedon commented:
This looks great! I don't have any major feedback, just what I've commented. Thanks for looking into this!

michaelDCurran committed:
Indent case bodies.

michaelDCurran committed:
Test names should start with uppercase
Also prepended all the resolve exception test names with 'ResolveThrows' for easier locating, and so there are no compile clashes with the actual exception classes.

michaelDCurran committed:
UiaOperationScope::ResolveInternal now returns status and ExctendedError in a pair, rather than by reference arguments.

michaelDCurran committed:
Add MalformedBytecodeException test for Uia operations abstraction API.

michaelDCurran commented:
@beweedon  I believe I have addressed all your comments. I settled on 200000 for the for loop. As 100000 is not enough and still throws InstructionLimitExceeded rather than MalformedBytecodeException.

csauls commented on src/UIAutomation/UiaOperationAbstraction/UiaOperationAbstraction.cpp:
```diff
@@ -2316,22 +2316,33 @@ namespace UiaOperationAbstraction
-        HRESULT operationResult = S_OK;
-        ResolveHrInternal(operationResult);
-        THROW_IF_FAILED(operationResult);
+                auto [status, extendedError] = ResolveInternal();
```
nit: This looks extra indented in my view.

csauls commented on src/UIAutomation/UiaOperationAbstraction/UiaOperationAbstraction.cpp:
```diff
@@ -2346,9 +2357,25 @@ namespace UiaOperationAbstraction
-                if (SUCCEEDED(operationResult))
+                // Fetch bound results on success, but also
+                // instruction limit exceeded, 
+                // As we no certainly some of the remote operation did execute.
```
nit: "no" -> "know"

csauls commented:
When the instruction limit is hit and partial results are returned, is there any risk that some of the bound results will be in an undefined state? Can the caller always inspect the values and determine if they are meaningful?

I understand from the description that it is up to the author to devise the method for resuming the operation, such as returning data that indicates how far the execution got before it was halted. Am I understanding correctly?

michaelDCurran committed:
Fix indenting

michaelDCurran committed:
Fix typo

michaelDCurran commented:
@csauls  I am under the assumption that all Uia types are initialized to a safe value. Thus resolving them at any point in the algorithm should be safe. In deed the low-level vm allows fetching these results via the resultSet for instructionLimitExceeded, but not UnhandledException for example it seems. Which suggests that there was a decision internally to allow fetching of partial results for instructionLimitExceeded, but not other errors.
I have no further knowledge here.
But yes, you are right that it will be up to the author to write their algorithm in such a way that allows resuming.
For example, if walking a very large text range (say a large MS Word document) to collect all headings, the author would keep shrinking the given textRange as they walk, so that restarting the algorithm, they will start from where they left off.
Of course this means that the textRange variable will have to be both bound as input and also bound for resolving.

csauls commented:
> @csauls I am under the assumption that all Uia types are initialized to a safe value. Thus resolving them at any point in the algorithm should be safe. In deed the low-level vm allows fetching these results via the resultSet for instructionLimitExceeded, but not UnhandledException for example it seems. Which suggests that there was a decision internally to allow fetching of partial results for instructionLimitExceeded, but not other errors. I have no further knowledge here. But yes, you are right that it will be up to the author to write their algorithm in such a way that allows resuming. For example, if walking a very large text range (say a large MS Word document) to collect all headings, the author would keep shrinking the given textRange as they walk, so that restarting the algorithm, they will start from where they left off. Of course this means that the textRange variable will have to be both bound as input and also bound for resolving.

Thanks for the additional details! I figured the objects would be in a safe state, I just didn't know if they'd be in a state to know if they had a meaningful value. A pointer may initialize to null, but there may not be a way to know if that's a result of the algorithm as well. Just something to think on if we need to have a way to make such distinctions in the future.

beweedon commented:
I did some more digging. You should actually be able to recover any output on any error, not just `InstructionLimitExceeded`, if you're using the low level bytecode API or `Microsoft.UI.UIAutomation`. It's just `UiaOperationAbstraction` that enforces not providing returned values on error.

This doesn't necessarily apply to this PR, but it's something to keep it mind I guess.

beweedon reviewed (approved):
Looks great to me! I'll merge this and the other one next week to give a bit more time for review!

beweedon commented:
/azp run

michaelDCurran commented:
I'm having trouble writing the MalformedBytecodeException test as it 
seems an unknown c++ exception is thrown from somewhere in either 
Serialize or AutomationRemoteOperation::execute, and I don't know what 
it is.
I will keep searching.
